sdk/src/main/scala/io/horizen/account/sc2sc/AbstractCrossChainRedeemMessageProcessor.scala

```diff
@@ -0,0 +1,150 @@
+      // validate arguments' semantics
+      validateSemantics(accountCcMsg.receiver, accountCcMsg.receiverSidechain, accountCcMsg.sender)
+
+      // Validate the receiving sidechain matches with this scId
+      validateScId(networkParams.sidechainId, accountCcMsg.receiverSidechain)
+
+      // Validate message has not been redeemed yet
+      val ccMsg = AbstractCrossChainMessageProcessor.buildCrosschainMessageFromAccount(accountCcMsg, networkParams)
+      validateDoubleMessageRedeem(ccMsg, view)
+
+      // Validate scCommitmentTreeRoot and nextScCommitmentTreeRoot exists
+      validateScCommitmentTreeRoot(ccRedeemMgs.scCommitmentTreeRoot, ccRedeemMgs.nextScCommitmentTreeRoot, view)
+
+      // Validate Proof
+      validateProof(ccRedeemMgs)
+    } catch {
+      case exception: IllegalArgumentException =>
+        throw new ExecutionRevertedException("Error while validating redeem message", exception)
+    }
+  }
+
+  private def validateSemantics(byteArraysToValidate: Array[Byte]*): Unit = {
+    val byteArrayMaxLength = 64
+    byteArraysToValidate.foreach(arrayToValidate => {
+      if (arrayToValidate.isEmpty || arrayToValidate.length > byteArrayMaxLength) {
+        throw new IllegalArgumentException(s"Value ${BytesUtils.toHexString(arrayToValidate)} is semantically wrong")
+      }
+    })
+  }
+
+  private def validateScId(scId: Array[Byte], receivingScId: Array[Byte]): Unit = {
+    if (!scId.sameElements(receivingScId)) {
+      throw new IllegalArgumentException(s"This scId `${BytesUtils.toHexString(scId)}` and receiving scId `${BytesUtils.toHexString(receivingScId)}` do not match")
+    }
+  }
+
+  private def validateDoubleMessageRedeem(ccMsg: CrossChainMessage, view: BaseAccountStateView): Unit = {
+    val currentMsgHash = sc2scCircuit.getCrossChainMessageHash(ccMsg)
+    val ccMsgFromRedeemAlreadyExists = view.doesCrossChainMessageHashFromRedeemMessageExist(currentMsgHash)
+    if (ccMsgFromRedeemAlreadyExists) {
+      throw new IllegalArgumentException(s"Message $ccMsg has already been redeemed")
+    }
+  }
+
+  private def validateScCommitmentTreeRoot(
+                                            scCommitmentTreeRoot: Array[Byte],
+                                            nextScCommitmentTreeRoot: Array[Byte],
+                                            view: BaseAccountStateView): Unit = {
+    if (!scTxMsgProc.doesScTxCommitmentTreeRootHashExist(scCommitmentTreeRoot, view)) {
+      throw new IllegalArgumentException(s"Sidechain commitment tree root `${BytesUtils.toHexString(scCommitmentTreeRoot)}` does not exist")
+    }
+
+    if (!scTxMsgProc.doesScTxCommitmentTreeRootHashExist(nextScCommitmentTreeRoot, view)) {
+      throw new IllegalArgumentException(s"Sidechain next commitment tree root `${BytesUtils.toHexString(nextScCommitmentTreeRoot)}` does not exist")
+    }
+  }
+
+  private def validateProof(ccRedeemMessage: AccountCrossChainRedeemMessage): Unit = {
```

Comment:
snark proof validation is very resource consuming operation, we need to evaluate the proper gas cost for it.
Otherwise, it can be a big bottleneck or even a source of attack.